File: packages/shared-ts/src/aggregate.ts
import type { ContributionDay } from './generated';
import { isoWeekKey, monthKey, yearKey } from './time';

export interface Bucket {
  key: string;
  total: number;
}

/** Sum of contributions for days in [from, to] inclusive. */
export function totalInWindow(days: ContributionDay[], from: string, to: string): number {
  let sum = 0;
  for (const day of days) {
    if (day.date >= from && day.date <= to) sum += day.contributionCount;
  }
  return sum;
}

/** Last N days as a bare number array, oldest-first, zero-padded if the input is shorter. */
export function sparklineSeries(days: ContributionDay[], n: number): number[] {
  if (n <= 0) return [];
  const sorted = [...days].sort((a, b) => a.date.localeCompare(b.date));
  const tail = sorted.slice(-n).map((day) => day.contributionCount);
  if (tail.length >= n) return tail;
  return [...new Array(n - tail.length).fill(0), ...tail];
}

/**
 * Current consecutive-day commit streak. Walks backward from `today` or from
 * the newest day in `days`. A zero-contribution *today* does not reset the
 * streak — the streak resumes from yesterday, matching what a developer would
 * intuit looking at their graph at 10am.
 */
export function streak(days: ContributionDay[], today?: string): number {
  if (days.length === 0) return 0;
  const byDate = new Map(days.map((d) => [d.date, d.contributionCount]));
  const sortedDates = [...byDate.keys()].sort();
  const anchor = today ?? sortedDates[sortedDates.length - 1]!;
  let cursor = anchor;
  let count = 0;
  let allowSkipToday = true;
  while (byDate.has(cursor)) {
    const contributions = byDate.get(cursor)!;
    if (contributions > 0) {
      count += 1;
      allowSkipToday = false;
    } else if (allowSkipToday && cursor === anchor) {
      allowSkipToday = false;
    } else {
      break;
    }
    cursor = previousDate(cursor);
  }
  return count;
}

function previousDate(date: string): string {
  const d = new Date(`${date}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() - 1);
  return d.toISOString().slice(0, 10);
}

function bucketBy<Key>(days: ContributionDay[], keyOf: (date: string) => Key): Map<Key, number> {
  const buckets = new Map<Key, number>();
  for (const day of days) {
    const key = keyOf(day.date);
    buckets.set(key, (buckets.get(key) ?? 0) + day.contributionCount);
  }
  return buckets;
}

export function byWeek(days: ContributionDay[]): Bucket[] {
  const buckets = bucketBy(days, isoWeekKey);
  return [...buckets.entries()]
    .map(([key, total]) => ({ key, total }))
    .sort((a, b) => a.key.localeCompare(b.key));
}

export function byMonth(days: ContributionDay[]): Bucket[] {
  const buckets = bucketBy(days, monthKey);
  return [...buckets.entries()]
    .map(([key, total]) => ({ key, total }))
    .sort((a, b) => a.key.localeCompare(b.key));
}

export function byYear(days: ContributionDay[]): Bucket[] {
  const buckets = bucketBy(days, yearKey);
  return [...buckets.entries()]
    .map(([key, total]) => ({ key: String(key), total }))
    .sort((a, b) => a.key.localeCompare(b.key));
}

File: packages/shared-swift/Sources/CompgitCore/Aggregate.swift
import Foundation
import CompgitSchema

/// Pure aggregation helpers over ContributionDay[].
/// Mirrors `packages/shared-ts/src/aggregate.ts`.

public struct Bucket: Equatable, Sendable {
    public let key: String
    public let total: Int

    public init(key: String, total: Int) {
        self.key = key
        self.total = total
    }
}

public enum CompgitAggregate {
    public static func totalInWindow(_ days: [ContributionDay], from: String, to: String) -> Int {
        var sum = 0
        for day in days where day.date >= from && day.date <= to {
            sum += day.contributionCount
        }
        return sum
    }

    /// Last `n` days, chronological, left-padded with zeros if the history is shorter.
    public static func sparklineSeries(_ days: [ContributionDay], n: Int) -> [Int] {
        guard n > 0 else { return [] }
        let sorted = days.sorted { $0.date < $1.date }
        let tail = sorted.suffix(n).map(\.contributionCount)
        if tail.count >= n { return Array(tail) }
        return Array(repeating: 0, count: n - tail.count) + tail
    }

    /// Current consecutive-day commit streak. A zero-contribution anchor day does
    /// NOT reset the streak — it resumes from the previous day.
    public static func streak(_ days: [ContributionDay], today: String? = nil) -> Int {
        guard !days.isEmpty else { return 0 }
        var byDate: [String: Int] = [:]
        for d in days { byDate[d.date] = d.contributionCount }
        let sortedDates = byDate.keys.sorted()
        let anchor = today ?? sortedDates.last!
        var cursor = anchor
        var count = 0
        var allowSkipToday = true
        while let contributions = byDate[cursor] {
            if contributions > 0 {
                count += 1
                allowSkipToday = false
            } else if allowSkipToday && cursor == anchor {
                allowSkipToday = false
            } else {
                break
            }
            cursor = previousDate(cursor) ?? ""
        }
        return count
    }

    public static func byWeek(_ days: [ContributionDay]) -> [Bucket] {
        return bucket(days, keyOf: CompgitTime.isoWeekKey)
    }

    public static func byMonth(_ days: [ContributionDay]) -> [Bucket] {
        return bucket(days, keyOf: CompgitTime.monthKey)
    }

    public static func byYear(_ days: [ContributionDay]) -> [Bucket] {
        return bucket(days, keyOf: { String(CompgitTime.yearKey($0)) })
    }

    private static func bucket(
        _ days: [ContributionDay],
        keyOf: (String) -> String
    ) -> [Bucket] {
        var totals: [String: Int] = [:]
        for day in days {
            let key = keyOf(day.date)
            totals[key, default: 0] += day.contributionCount
        }
        return totals
            .map { Bucket(key: $0.key, total: $0.value) }
            .sorted { $0.key < $1.key }
    }

    private static func previousDate(_ ymd: String) -> String? {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.locale = Locale(identifier: "en_US_POSIX")
        guard let date = formatter.date(from: ymd) else { return nil }
        let prior = date.addingTimeInterval(-86_400)
        return formatter.string(from: prior)
    }
}
